tests/SecurityGroup.py

```diff
@@ -0,0 +1,127 @@
+            Utils.Print("add trans: {}".format(json.dumps(trans, indent=4, sort_keys=True)))
+
+        if removeAction:
+            Utils.Print("removing {} from the security group".format(removeAction))
+            trans = self.defaultNode.pushMessage(self.contractAccount.name, "remove", removeAction, "--permission eosio@active")
+            Utils.Print("remove trans: {}".format(json.dumps(trans, indent=4, sort_keys=True)))
+
+        self.publishProcessNum += 1
+        self.publishTrans = self.defaultNode.pushMessage(self.contractAccount.name, "publish", "[{}]".format(self.publishProcessNum), "--permission eosio@active")[1]
+        Utils.Print("publish action trans: {}".format(json.dumps(self.publishTrans, indent=4, sort_keys=True)))
+        return self.publishTrans
+
+    def verifyParticipantsTransactionFinalized(self, transId = None):
+        Utils.Print("Verify participants are in sync")
+        if transId is None:
+            transId = Node.getTransId(self.publishTrans)
+        for part in self.participants:
+            if part.waitForTransFinalization(transId) == None:
+                Utils.errorExit("Transaction: {}, never finalized".format(trans))
+
+    def verifyNonParticipants(self, transId = None):
+        Utils.Print("Verify non-participants don't receive blocks")
+        if transId is None:
+            transId = Node.getTransId(self.publishTrans)
+        publishBlock = self.defaultNode.getBlockIdByTransId(transId)
+        prodLib = self.defaultNode.getBlockNum(blockType=BlockType.lib)
+        waitForLib = prodLib + 3 * 12
+        if self.defaultNode.waitForBlock(waitForLib, blockType=BlockType.lib) == None:
+            Utils.errorExit("Producer did not advance lib the expected amount.  Starting lib: {}, exp lib: {}, actual state: {}".format(prodLib, waitForLib, self.defaultNode.getInfo()))
+        producerHead = self.defaultNode.getBlockNum()
+
+        for nonParticipant in self.nonParticipants:
+            nonParticipantPostLIB = nonParticipant.getBlockNum(blockType=BlockType.lib)
+            assert nonParticipantPostLIB < publishBlock, "Participants not in security group should not have advanced LIB to {}, but it has advanced to {}".format(publishBlock, nonParticipantPostLIB)
+            nonParticipantHead = nonParticipant.getBlockNum()
+            assert nonParticipantHead < producerHead, "Participants (that are not producers themselves) should not advance head to {}, but it has advanced to {}".format(producerHead, nonParticipantHead)
+
+    def verifySecurityGroup(self, publishTrans = None):
+        if publishTrans is None:
+            publishTrans = self.publishTrans
+        publishTransId = Node.getTransId(publishTrans)
+        self.verifyParticipantsTransactionFinalized(publishTransId)
+        self.verifyNonParticipants(publishTransId)
+
+    def addToSecurityGroup(self):
```

Comment:
this method looks confusing. its meaning looks like `moveHeadToSecurityGroup`. I suggest to rename it or add new member like this:
`moveToSecurityGroup(self, index=0)`